lib/transaction/payload/proupregtxpayload.js

```diff
@@ -0,0 +1,239 @@
+  }
+
+  return payloadBufferWriter.toBuffer();
+};
+
+/**
+ * Parses raw ProUpRegTxPayload payload
+ * @param {Buffer} rawPayload
+ * @return {ProUpRegTxPayload}
+ */
+ProUpRegTxPayload.fromBuffer = function (rawPayload) {
+  var payloadBufferReader = new BufferReader(rawPayload);
+  var payload = new ProUpRegTxPayload();
+  var signatureSize = 0;
+  payload.version = payloadBufferReader.readUInt16LE();
+  payload.proTXHash = payloadBufferReader.read(HASH_SIZE).reverse().toString('hex');
+  // payload.mode = payloadBufferReader.readUInt16LE();
+  payload.keyIdOperator = payloadBufferReader.read(CKEYID_SIZE).reverse().toString('hex');
+  payload.keyIdVoting = payloadBufferReader.read(CKEYID_SIZE).reverse().toString('hex');
+  var scriptPayoutSize = payloadBufferReader.readVarintNum();
+  payload.scriptPayout = payloadBufferReader.read(scriptPayoutSize).reverse().toString('hex');
+  payload.inputsHash = payloadBufferReader.read(HASH_SIZE).reverse().toString('hex');
+
+  if (!payloadBufferReader.finished()) {
+    signatureSize = payloadBufferReader.readVarintNum();
+  }
+
+  if (signatureSize > 0) {
+    payload.payloadSig = payloadBufferReader.read(signatureSize).toString('hex');
+  }
+
+  if (!payloadBufferReader.finished()) {
+    throw new Error('Failed to parse payload: raw payload is bigger than expected.');
+  }
+
+  payload.validate();
+  return payload;
+};
+
+/**
+ * Creates new instance of ProUpRegTxPayload payload from JSON
+ * @param {string|ProUpRegTransactionPayloadJSON} payloadJSON
+ * @return {ProUpRegTxPayload}
+ */
+ProUpRegTxPayload.fromJSON = function fromJSON(payloadJSON) {
+  return new ProUpRegTxPayload(payloadJSON);
+};
+
+/**
+ * Validates ProUpRegTxPayload payload
+ * @param {ProUpRegTransactionPayloadJSON} payload
+ * @return {boolean}
+ */
+ProUpRegTxPayload.validatePayloadJSON = function (payload) {
+  if (!payload) {
+    throw new Error('No Payload specified');
+  }
+
+  Preconditions.checkArgumentType(payload.version, 'number', 'version');
+  Preconditions.checkArgumentType(payload.proTXHash, 'string', 'proTXHash');
+  // Preconditions.checkArgumentType(payload.mode, 'number', 'mode');
+  Preconditions.checkArgumentType(payload.keyIdOperator, 'string', 'keyIdOperator');
+  Preconditions.checkArgumentType(payload.keyIdVoting, 'string', 'keyIdVoting');
+  // Preconditions.checkArgumentType(payload.scriptPayoutSize, 'number', 'scriptPayoutSize');
+  Preconditions.checkArgumentType(payload.scriptPayout, 'string', 'scriptPayout');
+  Preconditions.checkArgumentType(payload.inputsHash, 'string', 'inputsHash');
+  // Preconditions.checkArgumentType(payload.payloadSigSize, 'number', 'payloadSigSize');
+  Preconditions.checkArgumentType(payload.payloadSig, 'string', 'payloadSig');
+
+  Preconditions.checkArgument(isUnsignedInteger(payload.version), 'Expected version to be an unsigned integer');
+  // Preconditions.checkArgument(isUnsignedInteger(payload.mode), 'Expected mode to be an unsigned integer');
```

Comment:
2 checks for mode?  I know its commented but the first one must be removed?

Reply:
Was just going by how it was done in previous checks. You check if it’s a number and then you check if it’s the unsigned integer type. I suppose it can be considered redundant to have the first. But maybe there were other considerations known before my implementation.